dispatcher.go

```diff
@@ -117,24 +119,46 @@ func (d dispatcher) Channel(service string) transport.Channel {
 }
 
 func (d dispatcher) Start() error {
+	var (
+		mu               sync.Mutex
+		startedInbounds  []transport.Inbound
+		startedOutbounds []transport.Outbound
+	)
+
 	service := transport.ServiceDetail{
 		Name:     d.Name,
 		Registry: d,
 	}
 
 	startInbound := func(i transport.Inbound) func() error {
 		return func() error {
-			return i.Start(service, d.deps)
+			err := i.Start(service, d.deps)
+			if err != nil {
+				return err
+			}
+
+			mu.Lock()
+			startedInbounds = append(startedInbounds, i)
+			mu.Unlock()
+			return nil
 		}
 	}
 
 	startOutbound := func(o transport.Outbound) func() error {
 		return func() error {
-			return o.Start(d.deps)
+			err := o.Start(d.deps)
+			if err != nil {
+				return err
+			}
+
+			mu.Lock()
+			startedOutbounds = append(startedOutbounds, o)
+			mu.Unlock()
+			return nil
 		}
 	}
 
-	var wait sync.ErrorWaiter
+	var wait intsync.ErrorWaiter
```

Comment:
I would've preferred `isync` for "internal sync" but @breerly doesn't like that.